src/lib.rs

```diff
@@ -228,24 +247,19 @@ pub extern "C" fn rustls_client_config_builder_load_roots_from_file(
 #[no_mangle]
 pub extern "C" fn rustls_client_config_free(config: *const rustls_client_config) {
     ffi_panic_boundary_unit! {
-        unsafe {
-            if let Some(c) = (config as *const ClientConfig).as_ref() {
-                // To free the client_config, we reconstruct the Arc. It should have a refcount of 1,
-                // representing the C code's copy. When it drops, that refcount will go down to 0
-                // and the inner ClientConfig will be dropped.
-                let arc: Arc<ClientConfig> = Arc::from_raw(c);
-                let strong_count = Arc::strong_count(&arc);
-                if strong_count < 1 {
-                    eprintln!(
-                        "rustls_client_config_free: invariant failed: arc.strong_count was < 1: {}. \
-                        You must not free the same client_config multiple times.",
-                        strong_count
-                    );
-                }
-            } else {
-                eprintln!("rustls_client_config_free: config was NULL");
-            }
-        };
+        let config: &ClientConfig = checked_ptr_ref!(config, ClientConfig, ());
```

Comment:
Looking at this usage, I don't love that `checked_ptr_ref!` can `return` an error value and bail out of the function, because I don't think the control flow is obvious when I read this function. Since there's no `?` or other error handling or branching, my assumption if I were scanning this code, perhaps seeking the line that produced the error message I found in a log, is that `checked_ptr_ref!` can't fail. What if instead of either evaluating to a `T` value or returning some value, `checked_ptr_ref!` evaluated to `Result<T>`, and then you'd do:
`let config = checked_ptr_ref!(config, ClientConfig, ())?;`
Which I think makes it more obvious that there's a potential failure case on this line. If you end up having to reintroduce the `match` statement at each callsite, then I think that would defeat the point of macro-ing away the checked conversion, but if it's just an extra `?`, I think it's worth it for clarity.

Reply:
I would love to do this. `?` only works in functions that return Result or Option. So it wouldn't work in this function that returns unit.

It seems like it *might* be possible for the functions that return rustls_result to return `Option<rustls_result>`, and get Rust to represent `None` as rustls_result::Ok (not sure how I would go about this but would do some research).

Another tweak that would provide much smaller benefit but much cheaper: I could rename these to start with `try_`, to remind people of the `try` macro that predated `?`: https://doc.rust-lang.org/std/macro.try.html. But given that's deprecated, it will probably resonate with fewer people over time.